### PR TITLE
Handle IConstructorBodyOperation and IMethodBodyOperation in CFG

### PR DESCRIPTION
**Pull request by mavasani**
Also added corresponding overloads to SemanticModel.GetControlFlowGraph

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -1221,6 +1225,68 @@ private void VisitStatements(IEnumerable<IOperation> statements)
...
+            {
+                EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.Locals, locals: operation.Locals));
+            }
+
+            if (operation.Initializer != null)
+            {
+                VisitStatement(operation.Initializer);
+            }
+
+            if (haveLocals)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: This is not correct. Locals declared via declaration patterns in a constructor initializer _are_ visible in the body.

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: For example, this is completely valid code (under C# 7.3):
```C#
class C : Base
{
    C() : base(out var i)
    {
        O(i);
    }
    void O(int i)
    {
        i = 1;
    }
}

class Base
{
    protected Base(out int i)
    {
        i = 1;
    }
}
```

---
In reply to: [189345445](https://github.com/dotnet/roslyn/pull/26955#discussion_r189345445) [](ancestors = 189345445)

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Ah, I got confused by the doc comment: http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IConstructorBodyOperation%2520.cs,21. Let me adjust the implementation and test baselines.

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Please also add a test like the one above, that way cfg verifier will fail if in the future we break this.

---
In reply to: [189346487](https://github.com/dotnet/roslyn/pull/26955#discussion_r189346487) [](ancestors = 189346487)

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: @333fred I can add a test, but note that existing test `ConstructorBody_14` does guard this case. The fact that the declared local is used or not in the body is not relevant.

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Yes it is. @AlekseyTs ensured that the CFG Verifier will verify that all locals used are contained in the surrounding scope, and it would have failed the above test case.

---
In reply to: [189379223](https://github.com/dotnet/roslyn/pull/26955#discussion_r189379223) [](ancestors = 189379223)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -1221,6 +1225,69 @@ private void VisitStatements(IEnumerable<IOperation> statements)
...
+
+        public override IOperation VisitMethodBodyOperation(IMethodBodyOperation operation, int? captureIdForResult)
+        {
+            Debug.Assert(_currentStatement == operation);
+
+            VisitMethodBodyBaseOperation(operation);
+            return null;
+        }
+
+        private void VisitMethodBodyBaseOperation(IMethodBodyBaseOperation operation)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider as a local function.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs [hunk 3]
```diff
@@ -281,5 +432,606 @@ class C
...
+@"
+Block[B0] - Entry
+    Statements (0)
+    Next (Regular) Block[B1]
+Block[B1] - Exit
+    Predecessors: [B0]
+    Statements (0)
+");
+        }
     }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: Consider a test where the base call has control flow:

``` 
public C(int? i) : base (i ?? 42)
{

}
```

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs [hunk 4]
```diff
@@ -281,5 +432,606 @@ class C
...
         }
+
+        [CompilerTrait(CompilerFeature.Dataflow)]
+        [Fact]
+        public void ConstructorBody_09()
+        {
+            string source = @"
+class C
+{
+    public C() : base()
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: Is constructor chaining `: this()` going to be done in a separate PR?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: I was in two minds about whether such a test will be useful, as the actual rewrite doesn't touch or special case operations within the initializer or body/expression blocks. I think it is worthwhile to at least have one test with flow in initializer, I will add one.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs (hunk 4 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: That shouldn't affect the flow graph, its just a different invocation in the initializer. I will still cover it in the new test I plan to add for your above suggestion.

Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: Agree it shouldn't be meaningfully different. Still would like to have at least one test to ensure we exercise the code path.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: I think this could be meaningfully interesting to have a single case here. The transition between ctor, base, and body has historically been a troublesome area for the compiler. I've introduced bugs there at least twice 😄 

Also in this particular case the two statements are not linked as an array. It's linking two separate nodes on an `IOperation` tree. Seems meaningful to ensure we get the right hookup here.

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 5)

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 5]
```diff
@@ -1045,11 +1045,15 @@ protected static void VerifyFlowGraphForTest<TSyntaxNode>(CSharpCompilation comp
...
+            VerifyFlowGraph(compilation, syntaxNode, expectedFlowGraph);
+        }
 
-            Operations.ControlFlowGraph graph = ControlFlowGraphVerifier.GetControlFlowGraph(syntaxNode, model);
-            ControlFlowGraphVerifier.VerifyGraph(compilation, expectedFlowGraph, graph);
+        protected static void VerifyFlowGraph(CSharpCompilation compilation, SyntaxNode syntaxNode, string expectedFlowGraph)
+        {
+            var model = compilation.GetSemanticModel(syntaxNode.SyntaxTree);
+            ControlFlowGraph graph = ControlFlowGraphVerifier.GetControlFlowGraph(syntaxNode, model);
+            ControlFlowGraphVerifier.VerifyGraph(model.Compilation, expectedFlowGraph, graph);
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs]: >model.Compilation [](start = 49, length = 17)

```compilation```?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 6]
```diff
@@ -1221,6 +1225,69 @@ private void VisitStatements(IEnumerable<IOperation> statements)
...
+
+            if (operation.BlockBody != null)
+            {
+                VisitStatement(operation.BlockBody);
+
+                // Check for error case with non-null BlockBody and non-null ExpressionBody.
+                if (operation.ExpressionBody != null)
+                {
+                    // Link last block of visited BlockBody to the exit block.
+                    BasicBlock lastBlock = _blocks.Last();
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >BasicBlock lastBlock = _blocks.Last(); [](start = 20, length = 38)

We always link CurrentBasicBlock

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs [hunk 7]
```diff
@@ -150,8 +196,39 @@ public C() : base()
...
+              Instance Receiver: 
+                IInstanceReferenceOperation (ReferenceKind: ContainingTypeInstance) (OperationKind.InstanceReference, Type: System.Object, IsImplicit) (Syntax: ': base()')
+              Arguments(0)
+
+    Next (Throw) Block[null]
+        ILiteralOperation (OperationKind.Literal, Type: null, Constant: null) (Syntax: 'null')
+Block[B2] - Block [UnReachable]
+    Predecessors (0)
+    Statements (1)
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: >IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null') [](start = 8, length = 112)

Where does this come from?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs [hunk 8]
```diff
@@ -214,8 +307,32 @@ public C()
...
+    Next (Regular) Block[B1]
+Block[B1] - Block
+    Predecessors: [B0]
+    Statements (0)
+    Next (Throw) Block[null]
+        ILiteralOperation (OperationKind.Literal, Type: null, Constant: null) (Syntax: 'null')
+Block[B2] - Block [UnReachable]
+    Predecessors (0)
+    Statements (1)
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: >IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null') [](start = 8, length = 112)

Same issue, it looks like this is already handled in the previous basic block

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: I see what is going on. It looks like we need to adjust implementation of VisitExpressionStatement to return null if its child is a Throw operation, after visiting the child of course. Could you please make this small adjustment, since this affects baselines that you are adding?

---
In reply to: [189704520](https://github.com/dotnet/roslyn/pull/26955#discussion_r189704520) [](ancestors = 189704520)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs [hunk 9]
```diff
@@ -128,6 +170,431 @@ public void M()
...
+                IParameterReferenceOperation: i (OperationKind.ParameterReference, Type: System.Int32, IsInvalid) (Syntax: 'i')
+
+    Next (Regular) Block[B3]
+
+.erroneous body {R1}
+{
+    Block[B2] - Block [UnReachable]
+        Predecessors (0)
+        Statements (1)
+            IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid, IsImplicit) (Syntax: 'j = i')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs]: >IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid, IsImplicit) (Syntax: 'j = i') [](start = 12, length = 118)

I cannot tell two assignments apart, please use distinct syntax

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs [hunk 10]
```diff
@@ -128,6 +170,431 @@ public void M()
...
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithFlowAnalysisFeature);
+
+            compilation.VerifyDiagnostics();
+
+            var tree = compilation.SyntaxTrees.Single();
+            var node1 = tree.GetRoot().DescendantNodes().OfType<BaseMethodDeclarationSyntax>().Single();
+
+            compilation.VerifyOperationTree(node1, expectedOperationTree:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs]: >compilation.VerifyOperationTree(node1, expectedOperationTree: [](start = 12, length = 61)

Same suggestion to get rid of tree validation in new Control Flow tests.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs (hunk 7 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs]: I was going to come talk to you about the difference between expression bodied and block case, thanks for suggesting the correct approach.

---
In reply to: [189708626](https://github.com/dotnet/roslyn/pull/26955#discussion_r189708626) [](ancestors = 189708626,189704520)

**Comment by mavasani** [comment]
@AlekseyTs addressed feedback

---
In reply to: [390789537](https://github.com/dotnet/roslyn/pull/26955#issuecomment-390789537) [](ancestors = 390789537)

**Review by 333fred — APPROVED** [review]
LGTM (iteration 6)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 6)
